### PR TITLE
Handle out of order messages

### PR DESCRIPTION
**Pull request by ariqchowdhury**
- blocktree adds blocks, even if they don’t have a parent
- commit rule becomes consecutive blocks and path to root exists
- when block without parent is added, send request sync
- this allows a node to continue proposing and moving consensus forward, even if it skips

### CODE REVIEW

File: monad-executor/src/mock_swarm.rs [hunk 1]
```diff
@@ -28,6 +30,8 @@ pub trait Transformer<M: Message> {
     // TODO smallvec? resulting Vec will almost always be len 1
     fn transform(&mut self, message: LinkMessage<M>) -> Vec<(Duration, LinkMessage<M>)>;
 
+    fn accumulated(&self) -> Option<&Vec<LinkMessage<M>>>;
```
Comment (omegablitz) [inline monad-executor/src/mock_swarm.rs]: I think this `accumulated` function shouldn't be necessary

File: monad-state/src/lib.rs [hunk 2]
```diff
@@ -380,6 +380,11 @@ where
                         ConsensusCommand::Unschedule => {
                             cmds.push(Command::TimerCommand(TimerCommand::Unschedule))
                         }
+                        ConsensusCommand::RequestSync { blockid: _ } => {
+                            // TODO: respond with the Proposal matching the blockID.
```
Comment (omegablitz) [inline monad-state/src/lib.rs:422]: super nit: maybe worth adding a trace log warn here

File: monad-state/tests/base.rs [hunk 3]
```diff
@@ -98,3 +100,64 @@ pub fn run_nodes<T: Transformer<MM>>(
...
+        .expect("no filtered messages")
+        .clone();
+
+    let iter_direction = if reverse {
+        itertools::Either::Right(filtered_msgs.into_iter().rev())
+    } else {
+        itertools::Either::Left(filtered_msgs.into_iter())
+    };
+
+    for msg in iter_direction {
```
Comment (omegablitz) [inline monad-state/tests/base.rs]: I think all of this could get folded into a transformer

File: monad-executor/src/state.rs [hunk 4]
```diff
@@ -79,7 +79,7 @@ pub trait State: Sized {
 
 pub trait Message: Clone {
     type Event;
-    type Id: Eq + Hash + Clone;
+    type Id: Eq + Hash + Clone + std::fmt::Debug;
```
Comment (omegablitz) [inline monad-executor/src/state.rs]: nit: is this necessary? it's fine if so, but just checking that it's not just a leftover from debugging

File: monad-state/src/lib.rs [hunk 5]
```diff
@@ -609,16 +620,20 @@ where
...
+        if qc.info.ledger_commit.commit_state_hash.is_some()
+            && self
+                .pending_block_tree
+                .path_to_root(&qc.info.vote.parent_id)
+        {
             let blocks_to_commit = self
                 .pending_block_tree
                 .prune(&qc.info.vote.parent_id)
-                .unwrap();
+                .unwrap_or_else(|_| panic!("\n{:?}", self.pending_block_tree));
```
Comment (omegablitz) [inline monad-state/src/lib.rs:683]: super nit: leftover custom unwrap, or is this still useful for debug?

if we see it happening in practice, may be nicer to make it a log fatal

File: monad-state/tests/base.rs (hunk 3 above)
Comment (omegablitz) [reply monad-state/tests/base.rs]: Btw, once this logic is encapsulated in a transformer, this whole function may be redundant - we might be able to have a single `run_node` function which we plug different combinations of transformers into.

We can save that for the future though

File: monad-blocktree/src/blocktree.rs [hunk 6]
```diff
@@ -159,31 +160,62 @@ impl<T: SignatureCollection> BlockTree<T> {
...
+            let Some(parent_id) = &i.parent else {return false };
+            bid = parent_id;
+        }
+    }
+
+    pub fn has_parent(&self, qc: &QuorumCertificate<T>) -> bool {
+        self.tree.contains_key(&qc.info.vote.id)
+    }
+
     pub fn debug_print(&self) -> std::io::Result<()> {
```
Comment (michael-yxchen) [inline monad-blocktree/src/blocktree.rs:202]: nit: I think we would want to print all the blocks that don't trace back to the root. It sounds annoying to maintain all the "roots", so maybe we can delay until we need it.

**Pull request merged**